src/types/edge-spec.ts

```diff
@@ -1,18 +1,66 @@
-import type { EdgeSpecRouteFn, EdgeSpecRouteParams } from "./web-handler"
+import {
+  createEdgeSpecRequest,
+  type EdgeSpecRouteFn,
+  type EdgeSpecRouteParams,
+} from "./web-handler.js"
 
-export type EdgeSpecRouteMatcher = (pathname: string) => {
-  matchedRoute: string
-  routeParams: EdgeSpecRouteParams
-}
+import type { ReadonlyDeep } from "type-fest"
+
+export type EdgeSpecRouteMatcher = (pathname: string) =>
+  | {
+      matchedRoute: string
+      routeParams: EdgeSpecRouteParams
+    }
+  | undefined
+  | null
 
-export interface EdgeSpec {
+export type EdgeSpecRouteMap = Record<string, EdgeSpecRouteFn>
+
+export interface EdgeSpecOptions {
   routeMatcher: EdgeSpecRouteMatcher
-  routeMapWithHandlers: {
-    [route: string]: EdgeSpecRouteFn
-  }
+  routeMapWithHandlers: EdgeSpecRouteMap
+
+  handleModuleServiceRouteNotFound?: EdgeSpecRouteFn
+  handle404?: EdgeSpecRouteFn
 }
 
-export type EdgeSpecAdapter<ReturnValue = void> = (
+// make this deeply immutable to force usage through helper functions
+export type EdgeSpec = ReadonlyDeep<EdgeSpecOptions>
```

Comment:
I think calling this `EdgeSpec` is kinda confusing, since I thought the `{methods: ["GET"], jsonBody: z.object({})}` object was an edge spec too

maybe `EdgeSpecServer`?

Reply:
`EdgeSpecRouteBundle`